File: LinkedList/src/lecture3/CheckPalindrome.java
package lecture3;

public class CheckPalindrome {

      public static Node reverseLinkedList(Node head) {
            Node prev = null;
            Node curr = head;

            while (curr != null) {
                  Node next = curr.next;
                  curr.next = prev;
                  prev = curr;
                  curr = next;
            }
            return prev;
      }

      @SuppressWarnings("JavaExistingMethodCanBeUsed")
      public static boolean isPalindrome(Node head) {
            if (head == null || head.next == null) return true;

            Node slow = head, fast = head;
            while (fast != null && fast.next != null) {
                  slow = slow.next;
                  fast = fast.next.next;
            }
            Node newHead = reverseLinkedList(slow.next);
            Node first = head, second = newHead;
            while (second != null) {
                  if (second.value != first.value) {
                        reverseLinkedList(newHead);
                        return false;
                  }
                  first = first.next;
                  second = second.next;
            }
            reverseLinkedList(newHead);
            return true;
      }
}

File: LinkedList_Go/lecture_3/check_palindrome.go
package main

func isPalindrome(head *ListNode) bool {
	if head == nil || head.Next == nil {
		return true
	}
	var fast, slow = head, head
	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	var secondHalf *ListNode
	if fast == nil {
		secondHalf = slow
	} else {
		secondHalf = slow.Next
	}
	secondHalf = reverseList(secondHalf)

	result := true
	first := head
	second := secondHalf

	for second != nil {
		if first.Val != second.Val {
			result = false
			break
		}
		first = first.Next
		second = second.Next
	}
	reverseList(secondHalf)
	return result
}
